### DISCUSSION TITLE
What are `xcode_install_storage_url` and `xcode_install_sas`?

### DISCUSSION THREAD

**Discussion by PrashantRaj18198** [actions/virtual-environments#9031] [General questions] [answered]
In macos templates, these two variables pop up but I can't find out how do I go about populating it.

**Comment by erik-bershel** [comment, marked as answer]
Hey @PrashantRaj18198!
That is the main thing. It's our production secrets. I may say that you may prepare your own strorage account (Azure or AWS based) and fill it with XCode installers (using `xcodes` for example).

**Reply by PrashantRaj18198** [reply]
Thanks for reaching out. By `xcodes` you mean https://github.com/XcodesOrg/xcodes right? And I need to download this and upload it to s3 (I'm using aws). I need to create a presigned url split the url on the `?` character, use the first part as the storage url and the latter as sas token.

**Reply by erik-bershel** [reply]
Very close. If you want to use our scripts without reworking, then you need to download all the necessary installers using the `xcodes` utility or from the Apple developers portal (doing this in runtime will be quite difficult due to the two-factor authorisation of the portal), then upload these installers to your SA. Our scripts expect interaction with Azure, so I'm afraid that with SAS tokens and links you will have to experiment a little and make some changes - I don't remember the AWS S3 standards very well.

**Reply by PrashantRaj18198** [reply]
Got it. Thanks.

**Reply by erik-bershel** [reply]
Glad to help. Feel free to ping me in case of other questions.

**Reply by HenAmar** [reply]
@PrashantRaj18198 can you please share if you were able to overcome this ?  Were you able to work with https://github.com/XcodesOrg/xcodes or have you downloaded the xcode for someplace else ? 
also, when placing in s3 or google storage, what were the values you put for xcode_install_storage_url and what were the values for xcode_install_sas ?

**Reply by PrashantRaj18198** [reply]
Yes. You need to download it to some place, the xcodes repo needs you apple developer username and password so you can't directly use it non-interactively. Place the downloaded file in an s3 bucket folder you need to follow the same naming convention for the files because the script looks for all the xcode download files with those versions under this folder path, the versions are present in the tools.json file for the macos runner. For example, `<bucket-name>/xcode-cache/Xcode-14.1.0-14B47b.xip` is the path for the 14.1.0 version of xcode. If you are using s3 beware that the default versions have `+` in them (something llke `Xcode-14.1.0+14B47b`, s3 url-encodes it's links so `+` is converted to `%2..` something which errors out the script. You'll need to get around it by replacing it to a character which is same after the url encode (I chose `-`). I tried with pre-signed link but that wasn't working for me, so I made the bucket public and left the sas token to be empty. So `xcode_install_storage_url: `<path-to-xcode-cache-dir>'` and `xcode_install_sas: ''`.

**Reply by deniojunior-hotmart** [reply]
Thanks for the suggestions @PrashantRaj18198, it helped me a lot!

I'd like to share with you the way I figured it out.

After some tests I discovered that S3 pre-signed URLs do not function properly due to the presence of special characters like `&`, which need to be escaped. The scripts currently used to construct the URL do not encode these characters, and I couldn't identify a workaround. Furthermore, pre-signed URLs are unique for each object, which poses an issue because the scripts require a consistent repository URL structure for downloading each Xcode object. I also discovered that the `s3://` protocol is not supported. 

In my situation, the S3 bucket policies did not allow the creation of public buckets. So to address this issue, I implemented a solution utilizing a private S3 bucket with CloudFront configured in front of it. This allows the objects to be accessed publicly via HTTPS.

**Reply by raghav155** [reply]
@erik-bershel Just curious, how does github manages to sync a new xcode version in their own storage ? Is this manual process or automated?

**Reply by erik-bershel** [reply]
Semiautomated due to the Apple limitations

**Comment by samus-aran** [comment]
We're locking this discussion because it has not had recent activity and/or other members have asked for more information to assist you but received no response. Thank you for helping us maintain a productive and tidy community for all our members.
